text/0034-ternary-hash.md

```diff
@@ -0,0 +1,415 @@
+
+    /// Reset the inner state of the sponge.
+    fn reset(&mut self);
+
+    /// Squeeze the sponge into a buffer
+    fn squeeze_into(&mut self, buf: &mut Trits) -> Result<(), Self::Error>;
+
+    /// Convenience function using `Sponge::squeeze_into` to to return an owned version of the hash.
+    fn squeeze(&mut self) -> Result<TritBuf, Self::Error> {
+        let mut output = TritBuf::zeros(Self::OUT_LEN);
+        self.squeeze_into(&mut output)?;
+        Ok(output)
+    }
+
+    /// Convenience function to absorb `input`, squeeze the sponge into a buffer, and reset the sponge in one go.
+    fn digest_into(&mut self, input: &Trits, buf: &mut Trits) -> Result<(), Self::Error> {
+        self.absorb(input)?;
+        self.squeeze_into(buf)?;
+        self.reset();
+        Ok(())
+    }
+
+    /// Convenience function to absorb `input`, squeeze the sponge, and reset the sponge in one go.
+    /// Returns an owned versin of the hash.
+    fn digest(&mut self, input: &Trits) -> Result<TritBuf, Self::Error> {
+        self.absorb(input)?;
+        let output = self.squeeze()?;
+        self.reset();
+        Ok(output)
+    }
+}
+```
+
+Following the sponge metaphor, an input reference provided by the user is `absorb`ed, and an output will be `squeeze`d
+from the data structure. `digest` is a convenience method calling `absorb` and `squeeze` in one go. The `*_into`
+versions of these methods are for passing a buffer into which the calculated hashes are written.  The internal state
+will not be cleared unless `reset` is called.
+
+### Design of `CurlP`
+
+`CurlP` is designed as a hash function that acts on a `T1B1` binary-encoded ternary buffer, with a hash length of `243`
+trits and an inner state of `729` trits, which we take as constants:
+
+```rust
+const HASH_LENGTH: usize = 243;
+const STATE_LENGTH: usize = HASH_LENGTH * 3;
+```
+
+In addition, a lookup table is used as part of the absorption step.
+2 is used to pad the table since there are only 9 combinations should be taken:
+
+```rust
+const TRUTH_TABLE: [i8; 11] = [1, 0, -1, 2, 1, -1, 0, 2, -1, 1, 0];
+```
+
+The way `CurlP` is defined, it can not actually fail, because the input or outputs can be of arbitrary size.
+Thus, the associated type `Error = !`, or `Error = Infallible` as `!` is not yet stabilized in Rust as of version
+`1.43`.
+
+This `enum` defines the number of rounds of hashing to apply before a hash is squeezed.
+```rust
+#[derive(Copy, Clone)]
+pub enum CurlPRounds {
+    Rounds27 = 27,
+    Rounds81 = 81,
+}
+```
+
+Type definition:
+
+```rust
+struct CurlP {
+    /// The number of rounds of hashing to apply before a hash is squeezed.
+    rounds: CurlPRounds,
+
+    /// The internal state.
+    state: TritBuf,
+
+    /// Workspace for performing transformations
+    work_state: TritBuf,
+}
+```
+
+In addition, there are two wrapper types for the very common `CurlP` variants with `27` and `81` rounds:
+
+```rust
+struct CurlP27(CurlP);
```

Comment:
Do these `impl Deref<Target=CurlP>`?

Reply:
Seems like they don't. They impl `Sponge` and forward the calls to the inner `CurlP`.

Reply:
Now they do.